Added Support for Extension-less Assets

bushrat011899 opened the pull request:
# Objective

- Addresses **Support processing and loading files without extensions** from #9714
- Addresses **More runtime loading configuration** from #9714
- Fixes #367
- Fixes #10703

## Solution

`AssetServer::load::<A>` and `AssetServer::load_with_settings::<A>` can now use the `Asset` type parameter `A` to select a registered `AssetLoader` without inspecting the provided `AssetPath`. This change cascades onto `LoadContext::load` and `LoadContext::load_with_settings`. This allows the loading of assets which have incorrect or ambiguous file extensions.

```rust
// Allow the type to be inferred by context
let handle = asset_server.load("data/asset_no_extension");

// Hint the type through the handle
let handle: Handle<CustomAsset> = asset_server.load("data/asset_no_extension");

// Explicit through turbofish
let handle = asset_server.load::<CustomAsset>("data/asset_no_extension");
```

Since a single `AssetPath` no longer maps 1:1 with an `Asset`, I've also modified how assets are loaded to permit multiple asset types to be loaded from a single path. This allows for two different `AssetLoaders` (which return different types of assets) to both load a single path (if requested).

```rust
// Uses GltfLoader
let model = asset_server.load::<Gltf>("cube.gltf");

// Hypothetical Blob loader for data transmission (for example)
let blob = asset_server.load::<Blob>("cube.gltf");
```

As these changes are reflected in the `LoadContext` as well as the `AssetServer`, custom `AssetLoaders` can also take advantage of this behaviour to create more complex assets.

---

## Change Log

- Updated `custom_asset` example to demonstrate extension-less assets.
- Added `AssetServer::get_handles_untyped` and Added `AssetServer::get_path_ids` 

## Notes

As a part of that refactor, I chose to store `AssetLoader`s (within `AssetLoaders`) using a `HashMap<TypeId, ...>` instead of a `Vec<...>`. My reasoning for this was I needed to add a relationship between `Asset` `TypeId`s and the `AssetLoader`, so instead of having a `Vec` and a `HashMap`, I combined the two, removing the `usize` index from the adjacent maps.

alice-i-cecile commented:
> I chose to refactor AssetLoaders into its own module because I was having trouble adding support for Asset TypeId information to it due to how tightly coupled it was to AssetServer. This wasn't strictly required, but I believe this is a better state for these types to be in, and should make it easier to update in the future.

I like this refactor! Decoupling that is a good step, and it's small enough that I don't feel terrible doing it in the same PR.

> Because I believe type information is of a higher priority than file extension, I have chosen to ignore the extension unless required (loading without supplying a TypeId). This allows for loading different types of assets from files with the same extension:

Oh this is *extremely* useful: I've run into this before and had to do bizarre `.foo.json` naming schemes. In the future, would we be able to get hints from metadata in the file itself? Definitely agree that type info > file extension though.

> As a part of that refactor, I chose to store AssetLoaders (within AssetLoaders) using a HashMap<TypeId, ...> instead of a Vec<...>. My reasoning for this was I needed to add a relationship between Asset TypeIds and the AssetLoader, so instead of having a Vec and a HashMap, I combined the two, removing the usize index from the adjacent maps.

Thanks for calling this out: it's a solid choice.

alice-i-cecile commented on assets/data/asset_no_extension:
```diff
@@ -0,0 +1,3 @@
+CustomAsset (
+    value: 13
+)
```
```suggestion
)

```

alice-i-cecile replied:
Then we don't get a complaint about a file not ending with a new line in the Github UI or elsewhere.

alice-i-cecile commented on crates/bevy_asset/src/server/loaders.rs:
```diff
@@ -0,0 +1,131 @@
+use std::{any::TypeId, sync::Arc};
+
+/// Storage for [`AssetLoader`]'s, providing helper methods for efficient access.
+#[derive(Default)]
```
```suggestion
#[derive(Debug, Default)]
```

alice-i-cecile commented on crates/bevy_asset/src/server/loaders.rs:
```diff
@@ -0,0 +1,131 @@
+}
+
+impl AssetLoaders {
+    pub fn get_by_asset_type_id(
```
Doc strings for public methods please.

alice-i-cecile commented on crates/bevy_asset/src/server/mod.rs:
```diff
@@ -963,6 +920,13 @@ pub struct MissingAssetLoaderForTypeNameError {
 }
 
+/// An error that occurs when an [`AssetLoader`] is not registered for a given [`Asset`] [`TypeId`].
+#[derive(Error, Debug)]
```
```suggestion
#[derive(Error, Debug, PartialEq)]
```

PartialEq is really helpful on public error types to let people write tests.

alice-i-cecile commented on crates/bevy_asset/src/server/mod.rs:
```diff
@@ -963,6 +920,13 @@ pub struct MissingAssetLoaderForTypeNameError {
+#[derive(Error, Debug)]
+#[error("no `AssetLoader` found with the ID '{type_id:?}'")]
+pub struct MissingAssetLoaderForTypeIdError {
+    type_id: TypeId,
```
```suggestion
    pub type_id: TypeId,
```

Might as well make things a bit more usable.

alice-i-cecile reviewed (changes requested):
The "disambiguate between assets of the same file extension via type annotations" bit of this PR is wildly useful: can you add / extend an example for it?

bushrat011899 commented:
> The "disambiguate between assets of the same file extension via type annotations" bit of this PR is wildly useful: can you add / extend an example for it?

I've updated the `custom_asset` example to demonstrate the creation of a `Blob` asset loader, which just simply reads the bytes of a file and stores them in a `Vec<u8>`:

```rust
#[derive(Asset, TypePath, Debug)]
pub struct Blob {
    pub bytes: Vec<u8>,
}
```

This allows loading the same path multiple times, provided they are for distinct asset types:

```rust
#[derive(Resource, Default)]
struct State {
    handle: Handle<CustomAsset>,
    other_handle: Handle<CustomAsset>,
    blob: Handle<Blob>,
    printed: bool,
}

fn setup(mut state: ResMut<State>, asset_server: Res<AssetServer>) {
    // Recommended way to load an asset
    state.handle = asset_server.load("data/asset.custom");

    // File extensions are optional, but are recommended
    state.other_handle = asset_server.load("data/asset_no_extension");

    // Will use BlobAssetLoader instead of CustomAssetLoader thanks to type inference
    state.blob = asset_server.load("data/asset.custom");
}
```

In the example, I also show that you can create a loader with no defined extensions:

```rust
impl AssetLoader for BlobAssetLoader {
    type Asset = Blob;
    type Settings = ();
    type Error = BlobAssetLoaderError;

    fn load<'a>(/* snip */) -> BoxedFuture<'a, Result<Self::Asset, Self::Error>> {
        /* snip */
    }

    fn extensions(&self) -> &[&str] {
        // This loader won't be used for any files by default
        &[]
    }
}
```

This blob loader I think is a good example of where you might want that behaviour; you have an asset which you want to load, but it's not the default way you would interact with that particular file. It would be possible to now provide a default implementation for `AssetLoader::extensions`, defaulting to this behaviour. However, I haven't done that, as I believe for most users most of the time, they do want to declare a file type (and extension) for their asset.

LennysLounge commented:
@alice-i-cecile bump for review.

alice-i-cecile commented:
So @bushrat011899, extending on my proposed use case a bit.

1. I have multiple types which should be stored as a .json file.
2. I want to parse the json file to determine what Rust type to load this in as, and then add it to a central collection.

Is that feasible with this PR?

That said, this example seems great to me, and I'm happy to approve this PR as is.

bushrat011899 commented:
> So @bushrat011899, extending on my proposed use case a bit.
> 
>     1. I have multiple types which should be stored as a .json file.
> 
>     2. I want to parse the json file to determine what Rust type to load this in as, and then add it to a central collection.
> 
> 
> Is that feasible with this PR?

@alice-i-cecile Yes that should be entirely possible:

1. First you'd create a plain JSON `AssetLoader` to give you the parsed contents to inspect.

```rust
// Handle<JsonAsset> implied by .json file extension if
// `JsonAssetLoader` is configured to be default handler of .json
let my_data_json_handle = asset_server.load("data/my_data.json");
```

2. Once loaded, you can then inspect the JSON data for whatever information you'd need to determine the type it *should* be:

```rust
let my_data_json = json_assets.get(&my_data_json_handle)?;

let serde_json::Value::String(magic_token) = my_data_json.get("magic_token")? else {
    return None;
};

// Can re-use the path value for DRY reasons
let path = my_data_json_handle.path().unwrap();

// All match arms need to have the same return type, using UntypedHandle to avoid an enum
let my_data_handle = match magic_token {
    "key_bindings" => asset_server.load::<KeyBindingsAsset>(&path).untyped(),
    "settings" => asset_server.load::<SettingsAsset>(&path).untyped(),
    "scoreboard" => asset_server.load::<ScoreboardAsset>(&path).untyped(),
    _ => my_data_json_handle.untyped(),
};

Some(my_data_handle)
```

I imagine you'd want an `enum` for the central collection, and that would probably be the type of `my_data_handle` in the above `match` statement instead of using `UntypedHandle`'s. Additionally, this behaviour is extended to the `LoadContext::load` method, so you could put all of this functionality into a custom `AssetLoader`. With #10565 and the `load_with_reader` method @nicopap suggested (would need another PR to add), you could even potentially reuse the `Reader`, reducing the number of IO operations. But that's all performance optimisation that is entirely optional.

alice-i-cecile commented:
Awesome, I'll spin out a follow-up issue for that.

alice-i-cecile commented:
@torsteingrindvik can I get your review here?

torsteingrindvik commented:
> @torsteingrindvik can I get your review here?

Yes- I have a use case for this so I'll try to test it before the week ends then I'll report back.

torsteingrindvik commented on crates/bevy_asset/src/server/loaders.rs:
```diff
@@ -0,0 +1,152 @@
+/// Storage for [`AssetLoader`]'s, providing helper methods for efficient access.
+#[derive(Debug, Default)]
+pub struct AssetLoaders {
+    values: HashMap<TypeId, MaybeAssetLoader>,
```
Could we find a more descriptive field name?

Names are hard, but perhaps

- `type_id_to_asset_loader`

or perhaps you have other ideas.

bushrat011899 replied:
No that's a good name, I just took the name from the original structure, back when that value was just a `Vec`

torsteingrindvik commented on crates/bevy_asset/src/server/loaders.rs:
```diff
@@ -0,0 +1,152 @@
+    /// Pre-register a loader that will later be added.
+    ///
+    /// Assets loaded with matching extensions will be blocked until the
+    /// real loader is added.
```
Could we hint users when pre-registration is necessary?

I remember at one point I stumbled upon a situation where I needed to register a loader at a specific part of Bevy's app init phase, and it took a while to figure out, so any hints are helpful (especially if it's related to that).

bushrat011899 replied:
I'm not too sure how to go about this, so I'm going to palm this off onto a future PR for now.

torsteingrindvik commented on crates/bevy_asset/src/server/mod.rs:
```diff
@@ -228,6 +192,22 @@ impl AssetServer {
     }
 
+    /// Retrieves the default [`AssetLoader`] for the given [`Asset`] [`TypeId`], if one can be found.
+    pub async fn get_asset_type_id_asset_loader<'a>(
```
Help me think: Is this more ergonomic for the user, or could something like

`pub async fn get_asset_asset_loader<'a, A: Asset>(&self) -> Result<...>`

work, where instead of providing the `TypeId` it's retrieved via the info impl'd on `Asset`?

bushrat011899 replied:
Fair point! I've renamed `get_asset_type_id_asset_loader` to `get_asset_loader_with_asset_type_id` and added `get_asset_loader_with_asset_type` which gets the `TypeId` for you. I think having the type-erased option is still important because it's easy to go from concrete -> erased, but not the other way around.

torsteingrindvik commented:
Just noting this early in my testing:

I switched to this branch in my codebase and now I see

```logs
2023-11-18T14:22:55.705371Z  WARN bevy_asset::server: Could not load asset via type_id
2023-11-18T14:22:55.705375Z  WARN bevy_asset::server: Could not load asset via type_id
2023-11-18T14:22:55.705393Z  WARN bevy_asset::server: Could not load asset via type_id
2023-11-18T14:22:55.705956Z  WARN bevy_asset::server: Could not load asset via type_id
2023-11-18T14:22:55.705968Z  WARN bevy_asset::server: Could not load asset via type_id
```

will investigate where these stem from, but noting that more informative logs would be good e.g. which type id is problematic, where did it try to load it from.

torsteingrindvik commented:
```logs
2023-11-18T16:41:36.048351Z ERROR bevy_asset::server: Requested handle of type TypeId { t: 299388583603764235311690958868876481710 } for asset 'foo.ron' does not match actual asset type 'path::to::MyBarAsset', which used loader 'path::to::MyBarAssetLoader'
```

I tried this scenario:

- Init an asset loader `MyBarAssetLoader` which loads `MyBarAsset` from ron files
- **Forget** on purpose to init asset loader `MyFooAssetLoader` which loads `MyFooAsset` from ron files
- Try to load a ron file which turns into `Handle<MyFooAsset>` from a ron file
- Get the above error

Two things:

1. If possible it would be great to map the verbose `TypeId` in the err msg into a human readable name of the type we're trying to load
2. Since this PR brings better asset type inference UX, it would be nice if the error message was more along the lines:

```logs
2023-11-18T16:41:36.048351Z ERROR bevy_asset::server:
Requested handle of asset type `MyFooAsset` for asset path 'foo.ron' does not have an asset loader.
Other asset loaders registered for this extension are: [
 'path::to::MyBarAssetLoader' for asset type 'path::to::MyBarAsset'
]
```

torsteingrindvik commented:
Happy with testing now, I can confirm the things I was hoping for is made possible by this PR.

This PR unlocks doing stuff like

```rust
app
  .init_asset_loader::<RonAssetLoader<Foo>>()
  .init_asset_loader::<RonAssetLoader<Bar>>()
  .init_asset_loader::<RonAssetLoader<Qux>>()
```

where `RonAssetLoader` has a blanket impl of `AssetLoader` if the generic is an asset.
Super ergonomic, thanks for this PR.

LennysLounge commented:
Would this not also make it possible to load assets without a path at all? 
For PR #10565 a new method was added to `LoadContext::load_direct_with_reader` to load an asset from a reader directly instead of from a file. Since the path parameter in that method was primairly used to indentify the correct loader, it would be nice if that can be eliminated.

However, there are some connections between an asset path and asset dependencies that i dont quiet understand yet.

bushrat011899 commented:
> Would this not also make it possible to load assets without a path at all?

Yes absolutely! I'll see how easy it would be to blend those changes into this PR for some even better ergonomics. If I can't get anything clean too quickly I'll palm it off for a future PR. But yes, definitely possible.

bushrat011899 commented:
> Just noting this early in my testing:
>
> ```
> 2023-11-18T14:22:55.705371Z  WARN bevy_asset::server: Could not load asset via type_id
> 2023-11-18T14:22:55.705375Z  WARN bevy_asset::server: Could not load asset via type_id
> 2023-11-18T14:22:55.705393Z  WARN bevy_asset::server: Could not load asset via type_id
> 2023-11-18T14:22:55.705956Z  WARN bevy_asset::server: Could not load asset via type_id
> 2023-11-18T14:22:55.705968Z  WARN bevy_asset::server: Could not load asset via type_id
> ```
> 
> EDIT: After changing the log message, they are all on the form
> 
> ```
> WARN bevy_asset::server: Could not load asset via type_id from <my asset folder absolute path>/models/Wall/scene.gltf.meta
> ```
> 
> likely since I don't use meta files for my gltf models. I haven't fully understood the processed/unprocessed workflow yet so not entirely sure if it's a problem on my end or not.

This is actually a really interesting error that I hadn't picked up on. The cause isn't because you're not using meta, it's because `GltfLoader` (which handles `.glb.` and `.gltf` files) can load `Scene`'s, but its `Asset` type is `Gltf`. The same behaviour can be seen when running the Alien Cake Addict example. For now, I've updated the `TypeId` resolution to just give up if there's a label provided with an `AssetPath`, since it could point to anything. In the future (Asset V2 maybe?) I hope that a better system could be used for contained/labelled assets which could register and expose this behaviour.

bushrat011899 commented:
> Would this not also make it possible to load assets without a path at all? For PR #10565 a new method was added to `LoadContext::load_direct_with_reader` to load an asset from a reader directly instead of from a file. Since the path parameter in that method was primairly used to indentify the correct loader, it would be nice if that can be eliminated.

I've updated this PR to provide/update 3 methods for working with `Reader`'s on the `LoadContext`:

### `load_direct_with_reader_and_path` (previously `load_direct_with_reader`)

Load an `ErasedLoadedAsset` using the provided `Reader`, claiming it was from a provided `Path`. This uses the extension within the provided path to determine what loader to use, and what underlying asset type is loaded. This would be most useful for `Zip` style loaders which might store a file path internally, and not know what type is actually stored internally.

```rust
for (path, reader) in parsed_zip_file.entries {
    let entry = load_context.load_direct_with_reader_and_path(&mut reader, path).await?;
    entries.push(entry);
}
```

### `load_direct_with_reader_and_type_id`

Load an `ErasedLoadedAsset` using the provided `Reader`, claiming it of a type matching the provided `TypeId`. This uses the `TypeId` provided to determine what loader to use, and what underlying asset type is loaded. This could be useful for runtime type checking, but I'm not certain how frequently it could be used.

### `load_direct_with_reader`

Load an `Asset` of type `A` using the provided `Reader`. This uses the type parameter `A` to determine what loader to use, and what underlying asset type is loaded. This is likely to be the preferred option for most users when the asset type can be known at compile time. This would be most useful for containers of known data types (e.g., the `GltfLoader` could use this to load `Scene`'s, etc.)

```rust
for (header, reader) in container.entries {
    match header.data_type {
        ContainedType::Scene => {
            let scene = load_context.load_direct_with_reader(&mut reader).await?;
            result.scenes.push(scene);
        }
        ContainedType::Mesh => {
            let mesh = load_context.load_direct_with_reader(&mut reader).await?;
            result.meshes.push(mesh);
        }
        // ...
    }
}
```

nicopap commented on crates/bevy_asset/src/loader.rs:
```diff
@@ -618,6 +618,97 @@ impl<'a> LoadContext<'a> {
+                    error,
+                },
+            })?
+            .take::<A>()
+            .ok_or_else(|| LoadDirectError {
```
Shouldn't this be an `unwrap`? Looks like by definition, a loader of type `AssetLoader` will always be registered for key `TypeId::of::<AssetLoader::Asset>()`

bushrat011899 replied:
Yes that is true! I personally avoid `unwrap` as much as possible (to a fault in this instance), which caused me to miss that fairly safe assumption.

nicopap commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -189,7 +189,11 @@ impl AssetInfos {
     ) -> Result<(UntypedHandle, bool), GetOrCreateHandleInternalError> {
-        match self.path_to_id.entry(path.clone()) {
+        // TODO: Need to re-work this to actually handle the None case
+        let type_id_unwrapped =
+            type_id.ok_or(GetOrCreateHandleInternalError::HandleMissingButTypeIdNotSpecified)?;
```
Looking at where this function gets used, there is only in `crates/bevy_asset/src/server/mod.rs` line 401 where there is a chance of it being `None`. There is a fairly large comment I didn't read which probably explains why someone would want to pass a `None` as `type_id` here.

bushrat011899 replied:
> This contains Some(UntypedHandle), if it was retrievable
> If it is None, that is because it was _not_ retrievable, due to
>
>    1. The handle was not already passed in for this path, meaning we can't just use that
>    2. The asset has not been loaded yet, meaning there is no existing Handle for it
>    3. The path has a label, meaning the AssetLoader's root asset type is not the path's asset type
>
> __In the None case, the only course of action is to wait for the asset to load so we can allocate the
> handle for that type.__

Since the `None` case for `type_id` can only occur when there is a label (`path.label().is_none().then(|| loader.asset_type_id())`), there's no way to infer the `TypeId` of the asset to load, so there's no way to generate a valid `Handle` to return back on `AssetInfos::get_or_create_path_handle_internal`. Based on how `unwrap_with_context` handles the `GetOrCreateHandleInternalError::HandleMissingButTypeIdNotSpecified` error case (no panic, just converts it to `None`), I think it's reasonable to early-return `Err` in `AssetInfos::get_or_create_path_handle_internal` when `type_id` is `None`. However, this probably needs a refactor at some point, since it doesn't feel right. To have a function always return `None` if a particular input is `None`. Feels like the input should be required at that point.

bushrat011899 replied:
I've updated the `TODO` comment to reflect this information, but I'm going to stop short of replacing it with a function that explicitly only takes `TypeId` just to limit scope.

nicopap commented on crates/bevy_asset/src/server/mod.rs:
```diff
@@ -728,10 +722,10 @@ impl AssetServer {
     /// Returns an active untyped handle for the given path, if the asset at the given path has already started loading,
     /// or is still "alive".
-    pub fn get_handle_untyped<'a>(&self, path: impl Into<AssetPath<'a>>) -> Option<UntypedHandle> {
+    pub fn get_handle_untyped<'a>(&self, path: impl Into<AssetPath<'a>>) -> Vec<UntypedHandle> {
```
```suggestion
    pub fn get_handles_untyped<'a>(&self, path: impl Into<AssetPath<'a>>) -> Vec<UntypedHandle> {
```

nicopap commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -264,9 +268,12 @@ impl AssetInfos {
-    pub(crate) fn get_path_handle(&self, path: AssetPath) -> Option<UntypedHandle> {
-        let id = *self.path_to_id.get(&path)?;
-        self.get_id_handle(id)
+    pub(crate) fn get_path_handles(&self, path: AssetPath) -> Vec<UntypedHandle> {
```
On a scale from 0 to 10, how impossible is it to return an `impl Iterator<Item=UntypedHandle> + '_`?

bushrat011899 replied:
For `AssetInfos::get_path_handles`? Easy.
For `AssetServer::get_handles_untyped`? I actually think impossible. It hides the `AssetInfo` behind a `RwLock`. So to return an `impl Iterator`, it needs to get a `RwLockGuard`, build the iterator, and then return both. The `RwLockGuard` can't outlive the iterator if it was included (e.g., a new struct owning the inner iterator and the guard), and there's nowhere sensible to store it.

If the `RwLock` was replaced with something with a little more control (e.g., somehow store a `&AssetInfos` only when the read lock should be held), but that's getting pretty involved pretty fast.

nicopap commented:
hmm, What about an additional field: `HashMap<AssetPath, Vec<TypeId>>` or `path_to_id: HashMap<AssetPath, Vec<UntypedAssetId>>`. After all `UntypedAssetId` already stores the `TypeId`.

nicopap commented on crates/bevy_asset/src/server/loaders.rs:
Nice initiative to isolate this `AssetLoaders` code.

bushrat011899 commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -264,9 +268,12 @@ impl AssetInfos {
+    pub(crate) fn get_path_handles(&self, path: AssetPath) -> Vec<UntypedHandle> {
+        self.path_to_id
+            .iter()
+            .filter(move |((key_path, _), _)| key_path == &path)
```
Agreed, I've updated `path_to_id` to be a `HashMap<AssetPath, HashMap<TypeId, UntypedAssetId>>`. This allows constant time access to all `UntypedAssetId`'s for a given path, and constant time access for a particular pair of asset path and type ID.

ricky26 commented on crates/bevy_asset/src/io/mod.rs:
```diff
@@ -248,10 +248,7 @@ impl AsyncRead for VecReader {
-        .to_os_string();
+    let mut extension = path.extension().unwrap_or_default().to_os_string();
     extension.push(".meta");
     meta_path.set_extension(extension);
```
nit: Could this be replaced with `meta_path.as_mut_os_string().push(b".meta");` now? Fishing out the extension seems like unnecessary work. (Would mean MSRV 1.70 though.)

bushrat011899 replied:
It could be, but I'm going to ignore since I'm uncertain what the status on the MSRV is right now, and I'd rather avoid the churn.

alice-i-cecile replied:
Bevy's MSRV policy is always "latest stable", but if things fail you'll want to bump the MSRV in Cargo.toml. Totally fine to ignore this for now.

ricky26 commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -189,7 +189,17 @@ impl AssetInfos {
     ) -> Result<(UntypedHandle, bool), GetOrCreateHandleInternalError> {
-        match self.path_to_id.entry(path.clone()) {
+        // TODO: It is not currently possible to create or get a `Handle` without the `type_id` parameter.
+        // This function should
```
TODO & suddenly ending sentence?

Should the behaviour here be to get any valid existing handle associated with the path?

bushrat011899 replied:
I've since resolved this TODO

NthTensor commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -189,7 +189,17 @@ impl AssetInfos {
-        match self.path_to_id.entry(path.clone()) {
+        // TODO: It is not currently possible to create or get a `Handle` without the `type_id` parameter.
+        // This function should
+        let type_id =
```
Can you explain why you moved this check out here, above the `weak_handle.upgrade()` branch? Is there a plan for resolving this?

bushrat011899 replied:
The check had to be moved as now multiple handles could be stored against a single path, whereas before a path (including label) was a unique address for an asset. As such, the `TypeId` is required to disambiguate which asset is desired from a given path.

To resolve this TODO, I've added an additional check to see if only 1 asset is currently registered against the path, and if so, work with that as the assumed correct one. Since the previous behaviour was to only ever have 1 or 0 assets per path, this should preserve the existing behaviour.

LennysLounge commented on crates/bevy_asset/src/loader.rs:
```diff
@@ -618,6 +618,90 @@ impl<'a> LoadContext<'a> {
+    /// For example, if you are deriving a new asset from the referenced asset, or you are building a collection of assets.
+    pub async fn load_direct_with_reader<'b, A: Asset>(
+        &mut self,
+        reader: &mut Reader<'_>,
+    ) -> Result<A, LoadDirectError> {
+        let loader = self
+            .asset_server
+            .get_asset_loader_with_asset_type::<A>()
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: error.into(),
+            })?;
+
+        let meta = loader.default_meta();
+
+        let load_context = LoadContext::new(
+            self.asset_server,
+            self.asset_path.clone_owned(),
+            self.should_load_dependencies,
+            self.populate_hashes,
+        );
+
+        let loaded_asset = loader
+            .load(reader, meta, load_context)
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: AssetLoadError::AssetLoaderError {
+                    path: self.asset_path.clone_owned(),
+                    loader_name: loader.type_name(),
+                    error,
+                },
+            })?
+            .take::<A>()
+            // AssetServer::get_asset_loader_with_asset_type::<A> can only return an `AssetLoader<Asset = A>`
+            .unwrap();
+
+        Ok(loaded_asset)
+    }
```
Doesnt this method do the exact same thing as `load_direct_with_reader_and_type_id` exect with the generic type instead? If that is the case, would it not make more sense to call `load_direct_with_reader_and_type_id` here and not duplicate this code?

Inside this method you call `asset_server.get_asset_loader_with_asset_type` which is just a wrapper around `asset_server.self.get_asset_loader_with_asset_type_id(TypeId::of::<A>())`. The whole method could be reduced to (not tested)
```rs
pub async fn load_direct_with_reader<'b, A: Asset>(reader..){
    self.load_direct_with_reader_and_type(reader, TypeId::of(A))
      .map(|loaded_asset| loaded_asset.take::<A>().unwrap())
}
```

bushrat011899 replied:
I've removed `load_direct_with_reader_and_type_id` as a part of upstreaming the current version of Bevy to get this PR into shape for merging (hopefully) soon. As a future PR, I'll look at adding them back in with appropriate application of DRY.

bushrat011899 commented on crates/bevy_asset/src/server/mod.rs:
```diff
@@ -728,10 +725,10 @@ impl AssetServer {
     /// Returns an active untyped handle for the given path, if the asset at the given path has already started loading,
     /// or is still "alive".
-    pub fn get_handle_untyped<'a>(&self, path: impl Into<AssetPath<'a>>) -> Option<UntypedHandle> {
+    pub fn get_handles_untyped<'a>(&self, path: impl Into<AssetPath<'a>>) -> Vec<UntypedHandle> {
```
Agreed. I've reverted the type signature of `get_handle_untyped` to return an option, with the value being the first item returned from `get_handles_untyped` (skipping the `Vec` collection). Likewise for path ID. I've updated comments for all 4 functions to make it clear what is happening and why.

bushrat011899 commented:
Sorry about the radio silence everyone, I had an unusually lovely holiday break over the new year, but I'm back now and wanting to sort out some of these outstanding PRs from last year.

I doubt this will be approved before the 0.13 merge window closes, but I've upstreamed this branch to the latest version of `main` and resolved all the merge conflicts that produced (of which there were enough that this was basically a full re-write! You guys were busy!)

I've also narrowed the scope of this PR to try and help spur its adoption:

1. There are no breaking changes in this PR now. All existing public APIs are unchanged in both signature and behaviour.
2. More specialised functions I added (like `load_direct_from_reader_with_type_id`) have been removed. I do think they are useful, but they aren't required to deliver on the key features of this PR.
3. I've removed the refactor I did (`loaders.rs`), since it nearly doubled the size of this PR on its own, and probably didn't help with the review process.

To focus on the core functionality of this PR:

1. Assets can now be loaded even if their path is missing an extension using type inference at the call site.
2. The same path can now be loaded as different asset types, again, via type inference at the loading call site.
3. `AssetLoader`'s no longer need to define extensions if they are not to be used as a default loader. Instead, they can omit the `extensions` method in the trait definition entirely.

Thanks again to everyone who reviewed this PR over the last couple months while I was away! ❤️

alice-i-cecile reviewed (approved):
Welcome back! I'm happy with this, and pleased to see that there's now no breaking changes.

NthTensor reviewed (commented):
Looks great, this is really high quality stuff! I didn't see any issues, but I am not submitting an approval because I still feel a little hazy about the asset server internals.

bushrat011899 commented on crates/bevy_asset/src/server/info.rs:
```diff
@@ -61,7 +61,7 @@ impl AssetInfo {
 #[derive(Default)]
 pub(crate) struct AssetInfos {
-    path_to_id: HashMap<AssetPath<'static>, UntypedAssetId>,
+    path_to_id: HashMap<AssetPath<'static>, HashMap<TypeId, UntypedAssetId>>,
```
Agreed, but I'll leave it off this PR, as the `TypeIdMap` type definition is private to the `bevy_ecs` crate. Ideally I think it would be moved into `bevy_utils`, since it has uses outside of the ECS crate.

torsteingrindvik commented on examples/asset/custom_asset.rs:
```diff
@@ -67,19 +107,55 @@ fn main() {
+    // Recommended way to load an asset
     state.handle = asset_server.load("data/asset.custom");
+
+    // File extensions are optional, but are recommended
```
Why are they recommended?

bushrat011899 replied:
Just a general tip regarding file management. Most file managers prefer file extensions as a way to determine default handlers. Specifically to Bevy, since file extensions are used as the last-resort for `Asset` disambiguation, it's good to try and have unique file extensions for asset types if possible. I could amend this comment to provide some explanation?

torsteingrindvik replied:
> since file extensions are used as the last-resort for Asset disambiguation

Yes I was hoping for a qualification like this such that users can know why something is recommended

torsteingrindvik commented:
Not sure if okay as-is or not, worth mentioning:

I tried having a `CustomAsset` and a `CustomAssetLoader` like your example, which works nicely.

However, if I try loading a `CustomAsset` from a file named e.g. `foo.ron`, I get an error that no `.ron` loader is added.
Loading it from `foo` (no extension) works.

So I suppose the question is why require listing the supported extension(s) in the loader trait impl at all, when type inference can help out?

torsteingrindvik commented:
I made an example which breaks and I think it shouldn't:


<details>
  <summary>Example</summary>

```ron
// my-asset
RonAsset (
    name: "Cube 1",
    translation: (0.1, 0.9, 2.4)
)
```

```ron
// my-asset-2
RonAsset2 (
    name: "Cube 2",
    translation: (0.8, 0.3, 0.9)
)
```

  ```rust
use bevy::{
    asset::{io::Reader, ron, AssetLoader, AsyncReadExt, LoadContext},
    prelude::*,
    reflect::TypePath,
    utils::{thiserror, thiserror::Error, BoxedFuture},
};
use serde::Deserialize;

#[derive(Asset, TypePath, Debug, Deserialize, Component)]
pub struct RonAsset {
    pub name: String,
    pub translation: Vec3,
}

#[derive(Asset, TypePath, Debug, Deserialize, Component)]
pub struct RonAsset2 {
    pub name: String,
    pub translation: Vec3,
}

#[derive(Default)]
pub struct RonLoader;

#[derive(Default)]
pub struct RonLoader2;

#[non_exhaustive]
#[derive(Debug, Error)]
pub enum RonLoaderError {
    /// An [IO](std::io) Error
    #[error("Could not load asset: {0}")]
    Io(#[from] std::io::Error),
    /// A [RON](ron) Error
    #[error("Could not parse RON: {0}")]
    RonSpannedError(#[from] ron::error::SpannedError),
}

impl AssetLoader for RonLoader {
    type Asset = RonAsset;
    type Settings = ();
    type Error = RonLoaderError;
    fn load<'a>(
        &'a self,
        reader: &'a mut Reader,
        _settings: &'a (),
        _load_context: &'a mut LoadContext,
    ) -> BoxedFuture<'a, Result<Self::Asset, Self::Error>> {
        Box::pin(async move {
            let mut bytes = Vec::new();
            reader.read_to_end(&mut bytes).await?;
            let custom_asset = ron::de::from_bytes::<RonAsset>(&bytes)?;
            Ok(custom_asset)
        })
    }
}

impl AssetLoader for RonLoader2 {
    type Asset = RonAsset2;
    type Settings = ();
    type Error = RonLoaderError;
    fn load<'a>(
        &'a self,
        reader: &'a mut Reader,
        _settings: &'a (),
        _load_context: &'a mut LoadContext,
    ) -> BoxedFuture<'a, Result<Self::Asset, Self::Error>> {
        Box::pin(async move {
            let mut bytes = Vec::new();
            reader.read_to_end(&mut bytes).await?;
            let custom_asset = ron::de::from_bytes::<RonAsset2>(&bytes)?;
            Ok(custom_asset)
        })
    }
}

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .init_asset::<RonAsset>()
        .init_asset_loader::<RonLoader>()
        .init_asset::<RonAsset2>()
        .init_asset_loader::<RonLoader2>()
        .add_systems(Startup, setup)
        .add_systems(Update, (set_cube_translation, set_cube_translation2))
        .run();
}

#[derive(Debug, Component)]
struct Cube;

fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
    asset_server: Res<AssetServer>,
) {
    // circular base
    commands.spawn(PbrBundle {
        mesh: meshes.add(shape::Circle::new(4.0)),
        material: materials.add(Color::WHITE),
        transform: Transform::from_rotation(Quat::from_rotation_x(-std::f32::consts::FRAC_PI_2)),
        ..default()
    });

    // cube 1
    commands.spawn((
        PbrBundle {
            mesh: meshes.add(shape::Cube { size: 1.0 }),
            material: materials.add(Color::rgb_u8(124, 144, 255)),
            ..default()
        },
        asset_server.load::<RonAsset>("data/my-asset"),
    ));

    // cube 2
    commands.spawn((
        PbrBundle {
            mesh: meshes.add(shape::Cube { size: 1.0 }),
            material: materials.add(Color::rgb_u8(124, 244, 255)),
            ..default()
        },
        asset_server.load::<RonAsset2>("data/my-asset-2"),
    ));

    // light
    commands.spawn(PointLightBundle {
        point_light: PointLight {
            intensity: 250_000.0,
            shadows_enabled: true,
            ..default()
        },
        transform: Transform::from_xyz(4.0, 8.0, 4.0),
        ..default()
    });
    // camera
    commands.spawn(Camera3dBundle {
        transform: Transform::from_xyz(-2.5, 4.5, 9.0).looking_at(Vec3::ZERO, Vec3::Y),
        ..default()
    });
}

fn set_cube_translation(
    mut transform_via_asset: Query<(&mut Transform, &Handle<RonAsset>)>,
    ron_assets: Res<Assets<RonAsset>>,
) {
    for (mut transform, asset_handle) in &mut transform_via_asset {
        let Some(asset) = ron_assets.get(asset_handle) else {
            continue;
        };
        if transform.translation != asset.translation {
            info!("Updating {}", asset.name);
            transform.translation = asset.translation;
        }
    }
}

fn set_cube_translation2(
    mut transform_via_asset: Query<(&mut Transform, &Handle<RonAsset2>)>,
    ron_assets: Res<Assets<RonAsset2>>,
) {
    for (mut transform, asset_handle) in &mut transform_via_asset {
        let Some(asset) = ron_assets.get(asset_handle) else {
            continue;
        };
        if transform.translation != asset.translation {
            info!("Updating {}", asset.name);
            transform.translation = asset.translation;
        }
    }
}

  ```
 
</details>


Run via `cargo run --example hot_asset_reloading --features="file_watcher"` (I modified the `hot_asset_reloading.rs` file).

Initially all seems ok:

```
2024-01-31T09:39:49.332644Z  INFO hot_asset_reloading: Updating Cube 1
2024-01-31T09:39:49.332663Z  INFO hot_asset_reloading: Updating Cube 2
```

Updating one of the files via hot-reloading gives an error:
```
2024-01-31T09:40:19.259971Z  INFO bevy_asset::server: Reloading data/my-asset because it has changed
2024-01-31T09:40:19.260257Z ERROR bevy_asset::server: no `AssetLoader` found for file with no extension
```

Updating the other file does not give an error, but the system updating the translation doesn't run like it should either
```
2024-01-31T09:40:29.994741Z  INFO bevy_asset::server: Reloading data/my-asset-2.ron because it has changed
```

bushrat011899 commented:
> I made an example which breaks and I think it shouldn't:
> Example
> 
> Run via `cargo run --example hot_asset_reloading --features="file_watcher"` (I modified the `hot_asset_reloading.rs` file).

Ah that's why I'm having trouble replicating this issue myself, I wasn't testing with the `file_watcher` feature. Thank you for posting this! I'll look at this and see if I can get this PR compatible with that feature as well.

bushrat011899 commented:
Ok @torsteingrindvik this should be resolved now. Again, thanks for testing this, I had completely forgotten to test against the `file_watcher` feature. As an aside, the fix involves now passing the original `input_handle` into the `load_internal` method called by `reload`, which should skip a fistful of redundant operations involved in getting/recreating the handle. Probably no performance impact in the real world, but still nice.

Running your example now works as expected, where editing the extensionless assets is immediately reflected in the running application with no runtime errors.

torsteingrindvik commented:
> Ok @torsteingrindvik this should be resolved now. Again, thanks for testing this, I had completely forgotten to test against the `file_watcher` feature. As an aside, the fix involves now passing the original `input_handle` into the `load_internal` method called by `reload`, which should skip a fistful of redundant operations involved in getting/recreating the handle. Probably no performance impact in the real world, but still nice.
> 
> Running your example now works as expected, where editing the extensionless assets is immediately reflected in the running application with no runtime errors.

Great! Approved. Thanks yourself for the nice PR.
